kernel/kernel-api/src/main/java/io/delta/kernel/internal/actions/Protocol.java

```diff
@@ -105,23 +140,216 @@ public Row toRow() {
+    Protocol protocolWithDependencies = withFeatures(feature.requiredFeatures());
+
+    if (feature.minReaderVersion() > protocolWithDependencies.minReaderVersion) {
+      throw new UnsupportedOperationException(
+          "TableFeature requires higher reader protocol version");
+    }
+
+    if (feature.minWriterVersion() > protocolWithDependencies.minWriterVersion) {
+      throw new UnsupportedOperationException(
+          "TableFeature requires higher writer protocol version");
+    }
+
+    boolean shouldAddToReaderFeatures =
+        feature.isReaderWriterFeature()
+            &&
+            // protocol already has support for `readerFeatures` set and the new feature
+            // can be explicitly added to the protocol's `readerFeatures`
+            supportsReaderFeatures();
+
+    boolean shouldAddToWriterFeatures = supportsWriterFeatures();
+
+    Set<String> newReaderFeatures = protocolWithDependencies.readerFeatures;
+    Set<String> newWriterFeatures = protocolWithDependencies.writerFeatures;
+
+    if (shouldAddToReaderFeatures) {
+      newReaderFeatures = new HashSet<>(protocolWithDependencies.readerFeatures);
+      newReaderFeatures.add(feature.featureName());
+    }
+
+    if (shouldAddToWriterFeatures) {
+      newWriterFeatures = new HashSet<>(protocolWithDependencies.writerFeatures);
+      newWriterFeatures.add(feature.featureName());
+    }
+
+    return new Protocol(
+        protocolWithDependencies.minReaderVersion,
+        protocolWithDependencies.minWriterVersion,
+        newReaderFeatures,
         newWriterFeatures);
   }
+
+  /**
+   * Determine whether this protocol can be safely upgraded to a new protocol `to`. This means: -
+   * all features supported by this protocol are supported by `to`.
+   *
+   * <p>Examples regarding feature status: - from `[appendOnly]` to `[appendOnly]` => allowed. -
+   * from `[appendOnly, changeDataFeed]` to `[appendOnly]` => not allowed. - from `[appendOnly]` to
+   * `[appendOnly, changeDataFeed]` => allowed.
+   */
+  public boolean canUpgradeTo(Protocol to) {
+    // All features supported by `this` are supported by `to`.
+    return to.getImplicitAndExplicitlyEnabledFeatures()
+        .containsAll(this.getImplicitAndExplicitlyEnabledFeatures());
+  }
+
+  /**
+   * Protocol normalization is the process of converting a table features protocol to the weakest
+   * possible form. This primarily refers to converting a table features protocol to a legacy
+   * protocol. A Table Features protocol can be represented with the legacy representation only when
+   * the features set of the former exactly matches a legacy protocol.
+   *
+   * <p>Normalization can also decrease the reader version of a table features protocol when it is
+   * higher than necessary.
+   *
+   * <p>For example: (1, 7, AppendOnly, Invariants, CheckConstraints) -> (1, 3) (3, 7, RowTracking)
+   * -> (1, 7, RowTracking)
+   */
+  public Protocol normalized() {
```

Comment:
is normalize the right/best word here? wasn't obvious to me. what about `minimized`?

Reply:
and then `maximized` for denormalized